tide/core/InactivityTimer.cpp

```diff
@@ -0,0 +1,107 @@
+/*      disclaimer in the documentation and/or other materials       */
+/*      provided with the distribution.                              */
+/*                                                                   */
+/*    THIS  SOFTWARE IS PROVIDED  BY THE  UNIVERSITY OF  TEXAS AT    */
+/*    AUSTIN  ``AS IS''  AND ANY  EXPRESS OR  IMPLIED WARRANTIES,    */
+/*    INCLUDING, BUT  NOT LIMITED  TO, THE IMPLIED  WARRANTIES OF    */
+/*    MERCHANTABILITY  AND FITNESS FOR  A PARTICULAR  PURPOSE ARE    */
+/*    DISCLAIMED.  IN  NO EVENT SHALL THE UNIVERSITY  OF TEXAS AT    */
+/*    AUSTIN OR CONTRIBUTORS BE  LIABLE FOR ANY DIRECT, INDIRECT,    */
+/*    INCIDENTAL,  SPECIAL, EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES    */
+/*    (INCLUDING, BUT  NOT LIMITED TO,  PROCUREMENT OF SUBSTITUTE    */
+/*    GOODS  OR  SERVICES; LOSS  OF  USE,  DATA,  OR PROFITS;  OR    */
+/*    BUSINESS INTERRUPTION) HOWEVER CAUSED  AND ON ANY THEORY OF    */
+/*    LIABILITY, WHETHER  IN CONTRACT, STRICT  LIABILITY, OR TORT    */
+/*    (INCLUDING NEGLIGENCE OR OTHERWISE)  ARISING IN ANY WAY OUT    */
+/*    OF  THE  USE OF  THIS  SOFTWARE,  EVEN  IF ADVISED  OF  THE    */
+/*    POSSIBILITY OF SUCH DAMAGE.                                    */
+/*                                                                   */
+/* The views and conclusions contained in the software and           */
+/* documentation are those of the authors and should not be          */
+/* interpreted as representing official policies, either expressed   */
+/* or implied, of Ecole polytechnique federale de Lausanne.          */
+/*********************************************************************/
+
+#include "InactivityTimer.h"
+
+#include "log.h"
+
+namespace
+{
+const int COUNTDOWNTIMER = 15000; // ms
+}
+
+InactivityTimer::InactivityTimer()
+{
+}
+
+InactivityTimer::InactivityTimer(const int timeout)
+    : _timeout(timeout * 60000) // from ms to minute
+    , _countDownTimer(new QTimer())
+    , _inactivityTimer(new QTimer())
+{
+    _inactivityTimer->setInterval(_timeout);
+    _inactivityTimer->setSingleShot(true);
+    _inactivityTimer->start();
+    _countDownTimer->setSingleShot(true);
+    _countDownTimer->setInterval(COUNTDOWNTIMER);
+
+    connect(_inactivityTimer.get(), &QTimer::timeout, [this]() {
+        if (!_countDownTimer->isActive())
+        {
+            _active = true;
+            _countDownTimer->start();
+            emit updated(shared_from_this());
+        }
+    });
+    connect(_countDownTimer.get(), &QTimer::timeout, [this]() {
+        emit poweroff();
+        _active = false;
+        emit updated(shared_from_this());
+    });
+}
+
+int InactivityTimer::getCountdownTimeout()
+{
+    return COUNTDOWNTIMER;
+}
+
+bool InactivityTimer::isActive()
+{
+    return _active;
+}
+
+void InactivityTimer::stop()
+{
+    _active = false;
+    _inactivityTimer->stop();
+    emit updated(shared_from_this());
```

Comment:
the updated signal should only be emitted when something really changes. Multiple calls to stop should do nothing.